### PR TITLE
Start work on removing .staging directory

### PR DESCRIPTION
**Pull request by anikitenko**
This PR created to change the way ".staging" directory was used.
In this PR:
* temp directory is determined based on OS you are running
* directory in "temp" one is named according to cons `OpsDir`
* file that contains information about contents in that directory is const `TMPContentsFile`
* another temp directories are created using Go `ioutil.TempDir` with prefix const `TMPStagingDirectoryPrefix`
* Uniqueness is done using the following methods:
  * If user uses `run` command then full path to it's binary is used (ex. `/Users/alexvwan/dev/src/test/server`) with it's association to new unique temp directory
  * If user uses `load` command then full path to current directory + gathered from manifest `manifest.Program`

### CODE REVIEW

**Comment by anikitenko** [comment]
So, in other words, instead of using ".staging" in each directory, now everything is located at: `<system temp dir>/<OpsDir>/*`. And information of existence of anything is located at `<system temp dir>/<OpsDir>/<TMPContentsFile>`

**Comment by anikitenko** [comment]
BTW, such logic of defining uniqueness is only valid until user don't specify parameters related to location of boot images.. I thought that if user specify any than responsibility exactly on users's side

**Comment by eyberg** [comment]
is there any reason besides debugging purposes that we even want to keep a .staging directory @tijoytom  && @mkhon ?? like - once the image is built imo we should just nuke it right? keep the resulting image but not the tmp stuff? and we could store the built image into something like ~/.ops/projects/my.img or something like that to help facilitate:

a) running it from anywhere (eg: outside the project root)

b) having a known good location to drop them in

also -- it doesn't need to be addressed in this pr but imo https://github.com/nanovms/ops/issues/196 prob. should go in the .staging (whatever becomes of this)

**Comment by mkhon** [comment]
Can we keep mkfs, boot.img and stage3.img in a fixed location (e.g.
somewhere under ~/.ops)?

**Comment by eyberg** [comment]
Id strongly prefer that as it can be reused

**Comment by anikitenko** [comment]
@mkhon I have open PR to keep mkfs, boot.img and stage3.img in a fixed location #195 but work in progress there..

File: lepton/const.go [hunk 1]
```diff
@@ -21,26 +21,34 @@ const manifest string = `(
...
+const TMPContentsFile string = "contents.json"
+const TMPStagingDirectoryPrefix string = "staging_"
+
 const ReleaseBaseUrl string = "https://storage.googleapis.com/nanos/release/%v"
 const DevBaseUrl string = "https://storage.googleapis.com/nanos/%v"
 
 const PackageBaseURL string = "https://storage.googleapis.com/packagehub/%v"
 const PackageManifestURL string = "https://storage.googleapis.com/packagehub/manifest.json"
 const PackageManifestFileName string = "manifest.json"
+const PackageRootPath string = "sysroot"
```
Comment (tijoytom) [inline lepton/const.go]: This should not be a global const if it's used only in one function.

File: lepton/image.go [hunk 2]
```diff
@@ -74,23 +74,36 @@ func addHostName(m *Manifest, c *Config) {
 }
 
 func addFilesFromPackage(packagepath string, m *Manifest) {
```
Comment (tijoytom) [inline lepton/image.go]: I don't think these changes work, i might be mistaken. Also there is no point in traversing the directory twice.

File: lepton/config.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ type Config struct {
 	RunConfig    RunConfig
 	Force        bool
 	TargetRoot   string
+	LocalTMPDir  string
```
Comment (tijoytom) [inline lepton/config.go]: TMP  should not be a config.

File: lepton/image.go [hunk 4]
```diff
@@ -204,17 +214,26 @@ func addMappedFiles(src string, dest string, m *Manifest) {
...
+
 	//  prepare manifest file
-	var elfmanifest string
+	elfmanifest := m.String()
 
-	elfmanifest = m.String()
 	// invoke mkfs to create the filesystem ie kernel + elf image
-	createFile(mergedImg)
-	args := []string{}
+	mergedImgPath := filepath.Join(c.LocalTMPDir, mergedImg)
```
Comment (tijoytom) [inline lepton/image.go]: Why is this necessary? creating and closing the temp file. Any way it's going to be created.

File: lepton/image.go [hunk 5]
```diff
@@ -230,19 +249,35 @@ func buildImage(c *Config, m *Manifest) error {
 		return errors.Wrap(err, 1)
 	}
 
+	bootImageOpen, err := os.Open(c.Boot)
```
Comment (tijoytom) [inline lepton/image.go]: Again why do you want to open this file?

File: lepton/image.go [hunk 6]
```diff
@@ -230,19 +249,35 @@ func buildImage(c *Config, m *Manifest) error {
...
+	if err != nil {
+		return errors.Wrap(err, 1)
+	}
+	defer bootImageOpen.Close()
+
+	mergedImgPathOpen, err := os.Open(mergedImgPath)
```
Comment (tijoytom) [inline lepton/image.go]: same as above

File: lepton/config.go (hunk 3 above)
Comment (anikitenko) [reply lepton/config.go]: Okay, I'll remove this from config

File: lepton/const.go (hunk 1 above)
Comment (anikitenko) [reply lepton/const.go]: Hm.. I thought it's better to move "sysroot" to const instead of hardcoding. If you think it's not needed, I'll remove it

File: lepton/image.go (hunk 2 above)
Comment (anikitenko) [reply lepton/image.go]: I tested these changes and they are working. Also I'm not traversing directory twice: first `filepath.Walk` traversing "sysroot", second one is it's parent directory ignoring "sysroot".

File: lepton/image.go (hunk 4 above)
Comment (anikitenko) [reply lepton/image.go]: I thought that there is no need to keep it open...

File: lepton/image.go (hunk 6 above)
Comment (anikitenko) [reply lepton/image.go]: `c.Boot` and `mergedImgPath` are need to be open to perform `io.Copy` to final image. I thought it's better to rewrite code to use `io.Copy` instead of running shell command `cat`

File: lepton/image.go (hunk 4 above)
Comment (anikitenko) [reply lepton/image.go]: ah, got what you mean.. I'll remove `mergedImg.Close()`

File: lepton/image.go (hunk 6 above)
Comment (tijoytom) [reply lepton/image.go]: is there a problem with cat? if not  i would keep it as such, using io.Copy buys you nothing, it introduce more code while solving a non-existing problem.

File: lepton/const.go (hunk 1 above)
Comment (tijoytom) [reply lepton/const.go]: I am fine if you want to keep it a const but then don't export it.(ie uppercase P)

File: lepton/image.go (hunk 6 above)
Comment (anikitenko) [reply lepton/image.go]: Actually there is no problem with `cat` but `io.Copy` is more a "Go way" of working with binary data..

File: lepton/const.go (hunk 1 above)
Comment (anikitenko) [reply lepton/const.go]: ah, okay, I'll make it not exported

File: lepton/image.go (hunk 6 above)
Comment (mkhon) [reply lepton/image.go]: I would prefer io.Copy as dependencies on external programs are bad

Comment (eyberg) [reply lepton/image.go]: yeh - it might be a bit slower but agreed - we should have as little exec.command as possible

**Comment by anikitenko** [comment]
I made changes based on discussions. I noticed that "image" is now placed in "$HOME/.ops/images" but it's name is hardcoded to "image". But what will happen if I run 2 images? I thought it's better to place it in temp directory determined by `GetTMPPathByProgramPath` func.. but for now I'm not exactly sure on how to do this.. :(

**Comment by anikitenko** [comment]
Just now an idea came to my mind.. when user has "shared" account on some server, some administrators can restrict access to `/tmp` directory.. What do you think if I rewrite `GetTMPPathByProgramPath` func to set path to staging like the following: `filepath.Join(GetOpsHome(), "staging")`? And the other functionality of func will remain? I think that this could resolve problem #196 and will implement in #210

**Comment by eyberg** [comment]
the tmp. image we don't want to keep around for anything other than debug purposes but for built images we need to store somewhere so it can be re-used again - in a known place - the images do need to be named in order for this to work but that wasn't clearly communicated - i suppose for now we can allow the old behavior by naming the image by default whatever directory it is in like go does but there should be a name argument && || spot in the config.json that will name it properly

sorry for the confusion

**Comment by anikitenko** [comment]
Ah, I see... work on that could be next step from moving from ".staging" directory.

**Comment by anikitenko** [comment]
Updated pull request to include fix of #217

**Comment by anikitenko** [comment]
Also in these changes I changed `os.TempDir` to `GetOpsHome()` in order to make sure that errors like #196 won't happen

And also in these changes I added funcs `RemoveTMPPathByProgramPath` and `RemoveImagePathByProgramPath` so they can be used to implement #147

File: lepton/image.go [hunk 7]
```diff
@@ -225,19 +241,43 @@ func buildImage(c *Config, m *Manifest) error {
...
 		return errors.Wrap(err, 1)
 	}
-	catcmd.Wait()
+
+	if err := mergedImg.Close(); err != nil {
+		return errors.Wrap(err, 1)
+	}
+
+	if len(c.Debugflags) == 0 {
+		if err := RemoveTMPPathByProgramPath(programAbsPath); err != nil {
```
Comment (tijoytom) [inline lepton/image.go]: Not sure i understand this, how is removing temp related to debug flags?

Comment (anikitenko) [reply lepton/image.go]: Ah, also forgot to mention that I added the following:
* If we have debug option specified than tempimage will remain
* If nothing debug specified that tempimage is removed along with it's temp directory at `$HOME/.ops/<temp_path>`

Comment (anikitenko) [reply lepton/image.go]: As @eyberg said that `tempimage` is more for debug purposes...

Comment (tijoytom) [reply lepton/image.go]: Sorry for the confusion, tempimage is an artifact of how we build the images. First we merge kernel an user executable to a file (which is the tempimage), then we merge tempimage and the boot loader to final image. So you can remove the temp-image right after the final merge.

Comment (anikitenko) [reply lepton/image.go]: I see.. but again, for debugging purposes maybe we can leave it if user specifies debug option?

Comment (eyberg) [reply lepton/image.go]: i don't see a need for tempimage or any of the other artifacts to be present -- '-d' works fine w/the resulting built image - for instance in the old model -- 'image' we want to keep but anything in ~/.staging we don't want to keep

Comment (anikitenko) [reply lepton/image.go]: Actually tempimage is not stored in .staging, everything is under ~/.ops, but Only in case someone specified -d option, tempimage is saved, in other cases it’s removed

Comment (eyberg) [reply lepton/image.go]: tempimage does not need to be saved at all -- '-d' only uses the built image - we can completely remove tempimage

Comment (anikitenko) [reply lepton/image.go]: Ah, okay, I’ll update PR to always remove it

Comment (anikitenko) [reply lepton/image.go]: Updated PR to always remove tempimage

File: lepton/images.go [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+		return nil
+	}
+
+	// Check if file which describes contents of images data exists
+	if _, err := os.Stat(imagesOpsFilePath); os.IsNotExist(err) {
+		return nil
+	}
+
+	// Read contents of file which describes contents of images data
+	contentJSON, err := ioutil.ReadFile(imagesOpsFilePath)
```
Comment (tijoytom) [inline lepton/images.go]: Not sure i understand why all this is necessary, are you storing the list of temp files in a json and using it to delete them later? Why can't you just delete the temp folder ie `os.RemoveAll(temp)` would delete everything.

**Comment by anikitenko** [comment]
Closing because PR is outdated
